File: packages/davinci/ios/Mapper/DaVinciNodeMapper.swift
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Foundation
import PingDavinci
import PingDavinciPlugin
import PingLogger
import PingOrchestrate
import RNPingCore

/// Maps native iOS DaVinci nodes and collectors to React Native bridge payloads.
enum DaVinciNodeMapper {
  private static let logTag = "DaVinciNodeMapper"

  /// Converts a native DaVinci node to a bridge-friendly dictionary payload.
  ///
  /// - Parameters:
  ///   - node: Native DaVinci node.
  ///   - logger: Optional Ping logger for non-fatal mapping warnings.
  /// - Returns: Serialized node payload.
  static func mapNode(_ node: Node, logger: Logger? = nil) -> NSDictionary {
    return JsonBridgeMapper.encodeJsonObject(mapNodePayload(node, logger: logger))
  }

  /// Builds a plain dictionary payload for a native node.
  ///
  /// - Parameters:
  ///   - node: Native DaVinci node.
  ///   - logger: Optional Ping logger for non-fatal mapping warnings.
  /// - Returns: Serializable node payload map.
  static func mapNodePayload(_ node: Node, logger: Logger? = nil) -> [String: Any] {
    var payload = [String: Any]()

    switch node {
    case let continueNode as ContinueNode:
      payload["type"] = "ContinueNode"
      payload["collectors"] = mapCollectors(continueNode, logger: logger)
      payload["input"] = JsonBridgeMapper.encodeJsonObject(continueNode.input)
      let unsupported = unsupportedFields(continueNode, logger: logger)
      if !unsupported.isEmpty {
        payload["unsupportedFields"] = unsupported
      }

    case let successNode as SuccessNode:
      payload["type"] = "SuccessNode"
      payload["session"] = ["value": successNode.session.value]

    case let errorNode as ErrorNode:
      payload["type"] = "ErrorNode"
      payload["message"] = errorNode.message
      payload["input"] = JsonBridgeMapper.encodeJsonObject(errorNode.input)
      if let status = errorNode.status {
        payload["status"] = status
      }

    case let failureNode as FailureNode:
      let causeMessage = failureNode.cause.localizedDescription
      payload["type"] = "FailureNode"
      payload["message"] = causeMessage
      payload["cause"] = causeMessage
      if case let ApiError.error(status, _, _) = failureNode.cause {
        payload["status"] = status
      }

    default:
      payload["type"] = "FailureNode"
      payload["message"] = "Unsupported node type: \(String(describing: type(of: node)))"
    }

    return payload
  }

  /// Serializes all `Collector` instances from a `ContinueNode` into an array.
  ///
  /// Unknown collector subtypes are skipped gracefully.
  ///
  /// - Parameters:
  ///   - node: Active continue node.
  ///   - logger: Optional Ping logger for non-fatal mapping warnings.
  /// - Returns: Serialized collectors array.
  private static func mapCollectors(_ node: ContinueNode, logger: Logger?) -> [[String: Any]] {
    var output = [[String: Any]]()
    for collector in node.collectors {
      if let map = mapCollector(collector, node: node, logger: logger) {
        output.append(map)
      }
    }
    return output
  }

  /// Diffs the raw form fields against the collectors the native SDK instantiated.
  ///
  /// Surfaces field entries from `node.input.form.components.fields[]` whose
  /// `inputType`/`type` was not registered with the native `CollectorFactory` — these
  /// are silently dropped by the SDK and would otherwise be invisible to JS consumers.
  ///
  /// The resolution order (`inputType` preferred over `type`) mirrors
  /// `CollectorFactory.collector(daVinci:from:)` in the iOS DaVinci SDK.
  ///
  /// - Parameters:
  ///   - node: Active continue node.
  ///   - logger: Optional Ping logger for non-fatal mapping warnings.
  /// - Returns: Array of `{ key, type }` entries describing dropped fields.
  private static func unsupportedFields(_ node: ContinueNode, logger: Logger?) -> [[String: Any]] {
    guard let fields = formFields(node, logger: logger) else { return [] }

    let registeredKeys = Set(node.collectors.map { $0.id })

    var entries = [[String: Any]]()
    for field in fields {
      guard let key = field["key"] as? String else { continue }

      // Mirrors CollectorFactory.collector(daVinci:from:): inputType takes precedence over type.
      guard let resolvedType = (field["inputType"] as? String) ?? (field["type"] as? String) else {
        continue
      }

      // A field is supported when the SDK instantiated a collector for its key.
      guard !registeredKeys.contains(key) else { continue }

      entries.append(["key": key, "type": resolvedType])
    }
    return entries
  }

  // MARK: - Form-field JSON navigation helpers

  /// Reads `node.input.form.components.fields[]` as a flat array of field dictionaries.
  ///
  /// - Parameters:
  ///   - node: Active continue node.
  ///   - logger: Optional Ping logger for non-fatal mapping warnings.
  /// - Returns: The fields array, or `nil` when the path is absent or malformed.
  private static func formFields(_ node: ContinueNode, logger: Logger?) -> [[String: Any]]? {
    guard
      let form = node.input["form"] as? [String: Any],
      let components = form["components"] as? [String: Any],
      let fields = components["fields"] as? [[String: Any]]
    else {
      return nil
    }
    return fields
  }

  /// Locates the field-level dictionary whose `key` matches `collectorKey`.
  ///
  /// - Parameters:
  ///   - collectorKey: The collector key to match against field entries.
  ///   - node: Active continue node.
  ///   - logger: Optional Ping logger for non-fatal mapping warnings.
  /// - Returns: The matching field dictionary, or `nil` when not present.
  private static func findFieldJson(
    _ collectorKey: String,
    node: ContinueNode,
    logger: Logger?
  ) -> [String: Any]? {
    return formFields(node, logger: logger)?.first { ($0["key"] as? String) == collectorKey }
  }

  /// Looks up the field-level JSON for a collector and emits it as `raw` on the collector map.
  ///
  /// Collectors whose key has no matching field entry (e.g. action buttons) have `raw` omitted.
  ///
  /// - Parameters:
  ///   - map: Mutable collector map to annotate with the raw field JSON.
  ///   - collectorKey: The collector key used to look up the matching field entry.
  ///   - node: Active continue node.
  ///   - logger: Optional Ping logger for non-fatal mapping warnings.
  private static func applyRawField(
    _ map: inout [String: Any],
    collectorKey: String,
    node: ContinueNode,
    logger: Logger?
  ) {
    guard let field = findFieldJson(collectorKey, node: node, logger: logger) else { return }
    map["raw"] = JsonBridgeMapper.encodeJsonObject(field)
  }

  /// Serializes a single collector to a bridge map.
  ///
  /// - Parameters:
  ///   - collector: Native collector instance to serialize.
  ///   - node: Active continue node providing form field context.
  ///   - logger: Optional Ping logger for non-fatal mapping warnings.
  /// - Returns: Serialized collector map, or `nil` for unknown collector subtypes.
  private static func mapCollector(
    _ collector: any Collector,
    node: ContinueNode,
    logger: Logger?
  ) -> [String: Any]? {
    var map: [String: Any]?
    switch collector {
    case let textCollector as TextCollector:
      map = mapTextCollector(textCollector)
    case let passwordCollector as PasswordCollector:
      map = mapPasswordCollector(passwordCollector, node: node, logger: logger)
    case let submitCollector as SubmitCollector:
      map = mapBaseFieldCollector(submitCollector)
    case let flowCollector as FlowCollector:
      map = mapBaseFieldCollector(flowCollector)
    case let labelCollector as LabelCollector:
      map = mapLabelCollector(labelCollector)
    case let singleSelectCollector as SingleSelectCollector:
      map = mapSingleSelectCollector(singleSelectCollector)
    case let multiSelectCollector as MultiSelectCollector:
      map = mapMultiSelectCollector(multiSelectCollector)
    case let phoneNumberCollector as PhoneNumberCollector:
      map = mapPhoneNumberCollector(phoneNumberCollector)
    case let registrationCollector as DeviceRegistrationCollector:
      map = mapDeviceRegistrationCollector(registrationCollector)
    case let authenticationCollector as DeviceAuthenticationCollector:
      map = mapDeviceAuthenticationCollector(authenticationCollector)
    default:
      logger?.w(
        "[\(logTag)] Skipping unsupported collector type: \(String(describing: type(of: collector)))",
        error: nil
      )
      return nil
    }
    guard var collectorMap = map else { return nil }
    applyRawField(&collectorMap, collectorKey: collector.id, node: node, logger: logger)
    return collectorMap
  }

  // MARK: - Field collector helpers

  /// Serializes base FieldCollector fields (key, type, label, required) into a map.
  ///
  /// - Parameter collector: FieldCollector instance.
  /// - Returns: Base collector map.
  private static func baseFieldCollectorMap<T>(_ collector: FieldCollector<T>) -> [String: Any] {
    return [
      "key": collector.key,
      "type": collector.type,
      "label": collector.label,
      "required": collector.required
    ]
  }

  /// Serializes base FieldCollector fields into a map without additional specialization.
  ///
  /// - Parameter collector: FieldCollector instance.
  /// - Returns: Base collector map.
  private static func mapBaseFieldCollector<T>(_ collector: FieldCollector<T>) -> [String: Any] {
    return baseFieldCollectorMap(collector)
  }

  /// Serializes a `TextCollector` to a bridge map, including value and optional validation regex.
  ///
  /// - Parameter collector: TextCollector instance.
  /// - Returns: Serialized text collector map.
  private static func mapTextCollector(_ collector: TextCollector) -> [String: Any] {
    var map = baseFieldCollectorMap(collector)
    map["value"] = collector.value
    if let validation = collector.validation, let regex = validation.regex {
      map["validation"] = ["regex": regex.pattern]
    }
    return map
  }

  /// Serializes a `PasswordCollector` to a bridge map.
  ///
  /// - Note: The `value` field is always emitted as `""` and never reflects the native
  ///   collector's current value. Password fields are treated as write-only across the
  ///   bridge — JS may push a value into the collector via `next()`, but the bridge
  ///   never round-trips the entered password back to JS. This matches the
  ///   `PasswordCollector.value` contract in the public TypeScript types and aligns
  ///   with the Android bridge's emission.
  private static func mapPasswordCollector(
    _ collector: PasswordCollector,
    node: ContinueNode,
    logger: Logger?
  ) -> [String: Any] {
    var map = baseFieldCollectorMap(collector)
    map["value"] = ""
    map["clearPassword"] = collector.clearPassword
    if let policy = extractPasswordPolicy(collector.key, node: node, logger: logger) {
      map["passwordPolicy"] = policy
    }
    return map
  }

  /// Reads passwordPolicy from the raw JSON at
  /// `continueNode.input.form.components.fields[].passwordPolicy` by matching the
  /// collector key.
  ///
  /// The SDK's own `passwordPolicy()` method reads from the node root
  /// (`continueNode.input["passwordPolicy"]`) which is the wrong path — that is why
  /// the bridge does the lookup itself.
  ///
  /// - Returns: Bridge-encoded password policy dictionary, or `nil` when missing.
  private static func extractPasswordPolicy(
    _ collectorKey: String,
    node: ContinueNode,
    logger: Logger?
  ) -> NSDictionary? {
    guard
      let field = findFieldJson(collectorKey, node: node, logger: logger),
      let policyDict = field["passwordPolicy"] as? [String: Any]
    else {
      return nil
    }

    do {
      let data = try JSONSerialization.data(withJSONObject: policyDict, options: [])
      let policy = try JSONDecoder().decode(PasswordPolicy.self, from: data)
      let encoded = try JSONEncoder().encode(policy)
      if let object = try JSONSerialization.jsonObject(with: encoded) as? [String: Any] {
        return JsonBridgeMapper.encodeJsonObject(object)
      }
      return nil
    } catch {
      logger?.w(
        "[\(logTag)] Failed to extract password policy for key=\(collectorKey)",
        error: error
      )
      return nil
    }
  }

  /// Serializes a `LabelCollector` to a bridge map.
  ///
  /// - Parameter collector: LabelCollector instance.
  /// - Returns: Serialized label collector map.
  private static func mapLabelCollector(_ collector: LabelCollector) -> [String: Any] {
    return [
      "key": collector.key,
      "type": "LABEL",
      "content": collector.content
    ]
  }

  /// Serializes a `SingleSelectCollector` to a bridge map, including value and options.
  ///
  /// - Parameter collector: SingleSelectCollector instance.
  /// - Returns: Serialized single-select collector map.
  private static func mapSingleSelectCollector(_ collector: SingleSelectCollector) -> [String: Any] {
    var map = baseFieldCollectorMap(collector)
    map["value"] = collector.value
    map["options"] = mapOptions(collector.options)
    if let validation = collector.validation, let regex = validation.regex {
      map["validation"] = ["regex": regex.pattern]
    }
    return map
  }

  /// Serializes a `MultiSelectCollector` to a bridge map, including value and options.
  ///
  /// - Parameter collector: MultiSelectCollector instance.
  /// - Returns: Serialized multi-select collector map.
  private static func mapMultiSelectCollector(_ collector: MultiSelectCollector) -> [String: Any] {
    var map = baseFieldCollectorMap(collector)
    map["value"] = collector.value
    map["options"] = mapOptions(collector.options)
    return map
  }

  /// Serializes a `PhoneNumberCollector` to a bridge map.
  ///
  /// - Parameter collector: PhoneNumberCollector instance.
  /// - Returns: Serialized phone number collector map.
  private static func mapPhoneNumberCollector(_ collector: PhoneNumberCollector) -> [String: Any] {
    var map = baseFieldCollectorMap(collector)
    map["defaultCountryCode"] = collector.defaultCountryCode
    map["validatePhoneNumber"] = collector.validatePhoneNumber
    map["countryCode"] = collector.countryCode
    map["phoneNumber"] = collector.phoneNumber
    return map
  }

  /// Serializes a `DeviceRegistrationCollector` to a bridge map, including available devices.
  ///
  /// - Parameter collector: DeviceRegistrationCollector instance.
  /// - Returns: Serialized device registration collector map.
  private static func mapDeviceRegistrationCollector(
    _ collector: DeviceRegistrationCollector
  ) -> [String: Any] {
    var map = baseFieldCollectorMap(collector)
    map["devices"] = mapDevices(collector.devices)
    return map
  }

  /// Serializes a `DeviceAuthenticationCollector` to a bridge map, including available devices.
  ///
  /// - Parameter collector: DeviceAuthenticationCollector instance.
  /// - Returns: Serialized device authentication collector map.
  private static func mapDeviceAuthenticationCollector(
    _ collector: DeviceAuthenticationCollector
  ) -> [String: Any] {
    var map = baseFieldCollectorMap(collector)
    map["devices"] = mapDevices(collector.devices)
    return map
  }

  /// Converts an array of `Option` values to serializable bridge maps.
  ///
  /// - Parameter options: Native option values from a select collector.
  /// - Returns: Array of `{ label, value }` dictionaries.
  private static func mapOptions(_ options: [Option]) -> [[String: Any]] {
    return options.map { option in
      ["label": option.label, "value": option.value]
    }
  }

  /// Converts an array of `Device` values to serializable bridge maps.
  ///
  /// - Parameter devices: Native device entries from a device collector.
  /// - Returns: Array of device dictionaries with type, title, iconSrc, isDefault, and optional id and description.
  private static func mapDevices(_ devices: [Device]) -> [[String: Any]] {
    return devices.map { device in
      var map: [String: Any] = [
        "type": device.type,
        "title": device.title,
        "iconSrc": device.iconSrc.absoluteString,
        "isDefault": device.isDefault ?? false
      ]
      if let id = device.id {
        map["id"] = id
      }
      if let description = device.description {
        map["description"] = description
      }
      return map
    }
  }
}

File: packages/davinci/android/src/main/java/com/pingidentity/rndavinci/mapper/DaVinciNodeMapper.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.rndavinci.mapper

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import kotlinx.serialization.json.Json
import com.pingidentity.davinci.collector.Device
import com.pingidentity.davinci.collector.DeviceAuthenticationCollector
import com.pingidentity.davinci.collector.DeviceRegistrationCollector
import com.pingidentity.davinci.collector.FlowCollector
import com.pingidentity.davinci.collector.LabelCollector
import com.pingidentity.davinci.collector.MultiSelectCollector
import com.pingidentity.davinci.collector.PasswordCollector
import com.pingidentity.davinci.collector.PasswordPolicy
import com.pingidentity.davinci.collector.PhoneNumberCollector
import com.pingidentity.davinci.collector.SingleSelectCollector
import com.pingidentity.davinci.collector.SubmitCollector
import com.pingidentity.davinci.collector.TextCollector
import com.pingidentity.davinci.plugin.Collector
import com.pingidentity.idp.davinci.IdpCollector
import com.pingidentity.logger.Logger
import com.pingidentity.orchestrate.ContinueNode
import com.pingidentity.orchestrate.ErrorNode
import com.pingidentity.orchestrate.FailureNode
import com.pingidentity.orchestrate.Node
import com.pingidentity.orchestrate.SuccessNode
import com.pingidentity.rncore.utils.JsonBridgeMapper
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Maps native DaVinci nodes and collectors to JS bridge payloads.
 */
internal object DaVinciNodeMapper {

    private const val TAG = "DaVinciNodeMapper"
    internal const val SOCIAL_LOGIN_BUTTON = "SOCIAL_LOGIN_BUTTON"

    private val json = Json { ignoreUnknownKeys = true }

    /**
     * Best-effort warning log through the configured Ping logger.
     */
    private fun logWarning(logger: Logger?, message: String, error: Throwable) {
        logger?.w("[$TAG] $message", error)
    }

    /**
     * Convert a native DaVinci node into a bridge-friendly map payload.
     *
     * @param node Native DaVinci node.
     * @param logger Optional Ping logger used for non-fatal mapping warnings.
     * @return Serialized node payload.
     */
    fun mapNode(node: Node, logger: Logger? = null): WritableMap {
        return Arguments.makeNativeMap(mapNodePayload(node, logger))
    }

    /**
     * Build a pure Kotlin payload for a native DaVinci node.
     *
     * @param node Native DaVinci node.
     * @param logger Optional Ping logger used for non-fatal mapping warnings.
     * @return Serializable node payload map.
     */
    fun mapNodePayload(node: Node, logger: Logger? = null): Map<String, Any?> {
        val payload = linkedMapOf<String, Any?>()

        when (node) {
            is ContinueNode -> {
                payload["type"] = "ContinueNode"
                payload["collectors"] = mapCollectorsPayload(node, logger)
                payload["input"] = JsonBridgeMapper.encodeJsonElement(node.input)
                val unsupported = unsupportedFieldsPayload(node, logger)
                if (unsupported.isNotEmpty()) {
                    payload["unsupportedFields"] = unsupported
                }
            }
            is SuccessNode -> {
                payload["type"] = "SuccessNode"
                payload["session"] = mapOf("value" to node.session.value)
            }
            is ErrorNode -> {
                payload["type"] = "ErrorNode"
                // TODO-SDK-PARITY: Android Orchestrate ErrorNode currently does not expose
                // an HTTP status field; iOS ErrorNode can include status metadata.
                payload["message"] = node.message
                payload["input"] = JsonBridgeMapper.encodeJsonElement(node.input)
            }
            is FailureNode -> {
                val causeMessage = node.cause.message ?: node.cause.toString()
                payload["type"] = "FailureNode"
                payload["message"] = causeMessage
                payload["cause"] = causeMessage
            }
        }

        return payload
    }

    /**
     * Serialize all `Collector<*>` instances from a `ContinueNode` into a list.
     * Unknown action types that do not implement `Collector` are skipped gracefully.
     *
     * @param node Active continue node.
     * @param logger Optional Ping logger used to surface skipped collector types.
     * @return Serialized collectors list.
     */
    private fun mapCollectorsPayload(node: ContinueNode, logger: Logger? = null): List<Map<String, Any?>> {
        return buildList {
            for (action in node.actions) {
                if (action !is Collector<*>) continue
                val collectorMap = mapCollectorPayload(action, node, logger) ?: continue
                add(collectorMap)
            }
        }
    }

    /**
     * Diff the raw form fields against the collectors the native SDK instantiated.
     *
     * Surfaces field entries from `node.input.form.components.fields[]` whose
     * `inputType`/`type` was not registered with the native `CollectorFactory`
     * — these are silently dropped by the SDK and would otherwise be invisible
     * to JS consumers.
     *
     * The resolution order (`inputType` preferred over `type`) mirrors
     * `CollectorFactory.collector(workflow, fields)` in the Android DaVinci SDK.
     *
     * @param node Active continue node.
     * @return List of `{ key, type }` entries describing dropped fields.
     */
    private fun unsupportedFieldsPayload(node: ContinueNode, logger: Logger? = null): List<Map<String, Any?>> {
        val fields = formFields(node, logger, "unsupportedFields diff") ?: return emptyList()

        val registeredKeys = node.actions
            .filterIsInstance<Collector<*>>()
            .mapNotNull { it.id() }
            .toSet()

        return buildList {
            for (element in fields) {
                val fieldJson = try {
                    element.jsonObject
                } catch (error: Exception) {
                    logWarning(logger, "Skipping non-object entry in form.components.fields[]", error)
                    continue
                }
                val key = fieldJson["key"]?.jsonPrimitive?.content ?: continue

                // Mirrors CollectorFactory.collector(): inputType takes precedence over type.
                val resolvedType = fieldJson["inputType"]?.jsonPrimitive?.content
                    ?: fieldJson["type"]?.jsonPrimitive?.content
                    ?: continue

                // SOCIAL_LOGIN_BUTTON fields are always handled via IdpCollector — exclude them.
                if (resolvedType == SOCIAL_LOGIN_BUTTON) continue

                // A field is supported when the SDK instantiated a collector for its key.
                if (registeredKeys.contains(key)) continue

                add(mapOf("key" to key, "type" to resolvedType))
            }
        }
    }

    /**
     * Serialize a single collector to a payload map.
     * Returns null when the collector type is unknown and should be skipped.
     *
     * @param collector Collector instance.
     * @param node Parent continue node (used for password policy extraction).
     * @param logger Optional Ping logger for non-fatal mapping warnings.
     * @return Serialized collector map, or null for unknown types.
     */
    private fun mapCollectorPayload(
        collector: Collector<*>,
        node: ContinueNode,
        logger: Logger? = null
    ): Map<String, Any?>? {
        val payload = when (collector) {
            is IdpCollector -> mapIdpCollector(collector)
            is TextCollector -> mapTextCollector(collector)
            is PasswordCollector -> mapPasswordCollector(collector, node, logger)
            is SubmitCollector -> mapBaseCollector(collector)
            is FlowCollector -> mapBaseCollector(collector)
            is LabelCollector -> mapLabelCollector(collector)
            is SingleSelectCollector -> mapSingleSelectCollector(collector)
            is MultiSelectCollector -> mapMultiSelectCollector(collector)
            is PhoneNumberCollector -> mapPhoneNumberCollector(collector)
            is DeviceRegistrationCollector -> mapDeviceRegistrationCollector(collector)
            is DeviceAuthenticationCollector -> mapDeviceAuthenticationCollector(collector)
            else -> {
                logWarning(
                    logger,
                    "Skipping unsupported collector type: ${collector::class.java.simpleName}",
                    UnsupportedOperationException("Unsupported collector: ${collector::class.java.name}")
                )
                null
            }
        } ?: return null
        val fieldKey = rawFieldKey(collector) ?: return null
        return applyRawField(payload, fieldKey, node, logger)
    }

    /**
     * Returns the stable raw-field key for a collector.
     *
     * Most collectors use [Collector.id] which equals the server `key` field. Collectors that
     * do not extend [FieldCollector] (e.g. [IdpCollector]) override with their own stable identifier.
     * Add a new branch here for any future collector whose id() does not match its form field key.
     */
    private fun rawFieldKey(collector: Collector<*>): String? = when (collector) {
        is IdpCollector -> collector.idpId
        else -> collector.id()
    }

    private fun mapIdpCollector(collector: IdpCollector): MutableMap<String, Any?> {
        val map = linkedMapOf<String, Any?>(
            "key" to collector.idpId,
            "type" to SOCIAL_LOGIN_BUTTON,
            "label" to collector.label,
            "idpId" to collector.idpId,
            "idpType" to collector.idpType,
            "idpEnabled" to collector.idpEnabled,
        )
        runCatching { collector.link.toString() }.getOrNull()?.let { map["link"] = it }
        return map
    }

    /**
     * Serialize base collector fields (key, type, label, required) into a map.
     */
    private fun baseCollectorMap(collector: com.pingidentity.davinci.collector.FieldCollector<*>): MutableMap<String, Any?> {
        return linkedMapOf(
            "key" to collector.key,
            "type" to collector.type,
            "label" to collector.label,
            "required" to collector.required
        )
    }

    private fun mapBaseCollector(collector: com.pingidentity.davinci.collector.FieldCollector<*>): Map<String, Any?> {
        return baseCollectorMap(collector)
    }

    private fun mapTextCollector(collector: TextCollector): Map<String, Any?> {
        val map = baseCollectorMap(collector)
        map["value"] = collector.value ?: ""
        collector.validation?.let { validation ->
            map["validation"] = mapOf("regex" to validation.regex.pattern)
        }
        return map
    }

    /**
     * Serialize a [PasswordCollector] to a payload map.
     *
     * @remarks
     * The `value` field is always emitted as `""` and never reflects the native collector's
     * current value. This is intentional: password fields are treated as write-only across
     * the bridge — JS may push a value into the collector via `next()`, but the bridge never
     * round-trips the entered password back to JS. This matches the `PasswordCollector.value`
     * contract in the public TypeScript types (`packages/davinci/src/types/node.types.ts`,
     * always `""` in bridge output) and aligns with the iOS bridge's emission.
     *
     * `clearPassword` controls whether the native collector wipes its in-memory value when
     * the parent node is closed; it does not affect the bridge value emission.
     *
     * @param collector Native password collector.
     * @param node Parent continue node (used to extract the field-level password policy from raw JSON).
     * @param logger Optional Ping logger for non-fatal policy-extraction warnings.
     * @return Serialized password collector map.
     */
    private fun mapPasswordCollector(
        collector: PasswordCollector,
        node: ContinueNode,
        logger: Logger? = null
    ): Map<String, Any?> {
        val map = baseCollectorMap(collector)
        map["value"] = ""
        map["clearPassword"] = collector.clearPassword
        extractPasswordPolicy(collector.key, node, logger)?.let { policy ->
            map["passwordPolicy"] = encodePasswordPolicy(policy)
        }
        return map
    }

    /**
     * Reads passwordPolicy from the raw JSON at
     * `continueNode.input.form.components.fields[].passwordPolicy` by matching the collector key.
     *
     * TODO-SDK-FUTURE-SUPPORT: The SDK's own `passwordPolicy()` method is `//TODO` in 2.0.1 because it reads from the
     * wrong JSON path (node root instead of the field-level object).
     */
    private fun extractPasswordPolicy(
        collectorKey: String,
        node: ContinueNode,
        logger: Logger? = null
    ): PasswordPolicy? {
        return try {
            val fieldJson = findFieldJson(node, collectorKey, logger) ?: return null
            val policyJson = fieldJson["passwordPolicy"]?.jsonObject ?: return null
            json.decodeFromJsonElement(PasswordPolicy.serializer(), policyJson)
        } catch (error: Exception) {
            logWarning(logger, "Failed to extract password policy for key=$collectorKey", error)
            null
        }
    }

    /**
     * Read `node.input.form.components.fields[]` as a JSON array.
     *
     * @param node Active continue node.
     * @param logger Optional Ping logger for non-fatal navigation warnings.
     * @param contextLabel Short identifier used in warning messages so callers are distinguishable.
     * @return The fields array, or null if the form/components/fields path is absent or malformed.
     */
    private fun formFields(
        node: ContinueNode,
        logger: Logger? = null,
        contextLabel: String = "form fields lookup"
    ): kotlinx.serialization.json.JsonArray? {
        return try {
            node.input
                .get("form")?.jsonObject
                ?.get("components")?.jsonObject
                ?.get("fields")?.jsonArray
        } catch (error: Exception) {
            logWarning(logger, "Failed to parse form fields for $contextLabel", error)
            null
        }
    }

    /**
     * Locate the field-level JSON object whose `key` matches the given collector key.
     *
     * @param node Active continue node.
     * @param collectorKey Collector key to look up under `form.components.fields[]`.
     * @param logger Optional Ping logger for non-fatal navigation warnings.
     * @return The matching field object, or null when not present.
     */
    private fun findFieldJson(
        node: ContinueNode,
        collectorKey: String,
        logger: Logger? = null
    ): JsonObject? {
        val fields = formFields(node, logger, "raw field lookup for key=$collectorKey") ?: return null
        return fields.firstOrNull {
            try {
                it.jsonObject["key"]?.jsonPrimitive?.content == collectorKey
            } catch (error: Exception) {
                logWarning(logger, "Skipping non-object entry while resolving raw field for key=$collectorKey", error)
                false
            }
        }?.jsonObject
    }

    /**
     * Look up the field-level JSON for a collector and emit it as `raw` on the collector map.
     * Collectors whose key has no matching field entry (e.g. action buttons) have `raw` omitted.
     */
    private fun applyRawField(
        payload: Map<String, Any?>,
        collectorKey: String,
        node: ContinueNode,
        logger: Logger? = null
    ): Map<String, Any?> {
        val fieldJson = findFieldJson(node, collectorKey, logger) ?: return payload
        val withRaw = LinkedHashMap(payload)
        withRaw["raw"] = JsonBridgeMapper.encodeJsonElement(fieldJson)
        return withRaw
    }

    private fun encodePasswordPolicy(policy: PasswordPolicy): Any? {
        val policyJson = json.encodeToJsonElement(PasswordPolicy.serializer(), policy)
        return JsonBridgeMapper.encodeJsonElement(policyJson.jsonObject)
    }

    private fun mapLabelCollector(collector: LabelCollector): Map<String, Any?> {
        return linkedMapOf(
            "key" to collector.key,
            "type" to "LABEL",
            "content" to collector.content
        )
    }

    private fun mapSingleSelectCollector(collector: SingleSelectCollector): Map<String, Any?> {
        val map = baseCollectorMap(collector)
        map["value"] = collector.value ?: ""
        map["options"] = mapOptions(collector.options)
        collector.validation?.let { validation ->
            map["validation"] = mapOf("regex" to validation.regex.pattern)
        }
        return map
    }

    private fun mapMultiSelectCollector(collector: MultiSelectCollector): Map<String, Any?> {
        val map = baseCollectorMap(collector)
        map["value"] = collector.value?.toList() ?: emptyList<String>()
        map["options"] = mapOptions(collector.options)
        return map
    }

    private fun mapPhoneNumberCollector(collector: PhoneNumberCollector): Map<String, Any?> {
        val map = baseCollectorMap(collector)
        map["defaultCountryCode"] = collector.defaultCountryCode
        map["validatePhoneNumber"] = collector.validatePhoneNumber
        map["countryCode"] = collector.countryCode
        map["phoneNumber"] = collector.phoneNumber
        return map
    }

    private fun mapDeviceRegistrationCollector(collector: DeviceRegistrationCollector): Map<String, Any?> {
        val map = baseCollectorMap(collector)
        map["devices"] = mapDevices(collector.devices)
        return map
    }

    private fun mapDeviceAuthenticationCollector(collector: DeviceAuthenticationCollector): Map<String, Any?> {
        val map = baseCollectorMap(collector)
        map["devices"] = mapDevices(collector.devices)
        return map
    }

    private fun mapOptions(options: List<com.pingidentity.davinci.collector.Option>): List<Map<String, Any?>> {
        return options.map { option ->
            mapOf(
                "label" to option.label,
                "value" to option.value
            )
        }
    }

    private fun mapDevices(devices: List<Device>): List<Map<String, Any?>> {
        return devices.map { device ->
            buildMap {
                device.id?.let { put("id", it) }
                put("type", device.type)
                put("title", device.title)
                device.description?.let { put("description", it) }
                put("iconSrc", device.iconSrc)
                put("isDefault", device.default)
            }
        }
    }
}
